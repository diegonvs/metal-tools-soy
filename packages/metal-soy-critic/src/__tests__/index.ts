import * as child_process from 'child_process';
import * as path from 'path';

describe('cli', () => {
  function runCliFrom(bin: string, cwd: string, ...args: Array<string>): Promise<[number, string]> {
    return new Promise((resolve, reject) => {
      const output: Array<string> = [];
      const child = child_process.fork(bin, args, {cwd, silent: true});

      child.stdout.on('data', data => {
        output.push(data.toString());
      });

      child.on('close', code => {
        resolve([code, output.join('')]);
      });
    });
  }

  function runCli(...args: Array<string>): Promise<[number, string]> {
    return runCliFrom('./lib/index.js', process.cwd(), ...args);
  }

  test('should print usage without args', async () => {
    const [exitCode, output] = await runCli();

    expect(exitCode).toBe(0);
    expect(output).toMatchSnapshot();
  });

  test('should validate file', async () => {
    const [exitCode, output] = await runCli('test/fixtures/MissingInternal.soy');

    expect(exitCode).toBe(1);
    expect(output).toMatchSnapshot();
  });

  test('should print status when no errors', async () => {
    const [exitCode, output] = await runCli('test/fixtures/Test.soy');

    expect(exitCode).toBe(0);
    expect(output).toMatchSnapshot();
  });

  test('should print more with verbose flag', async () => {
    const [exitCode, output] = await runCli('-v', 'test/fixtures/Test.soy');

    expect(exitCode).toBe(0);
    expect(output).toMatchSnapshot();
  });

  test('should recursively look for Soy files if passed a directory', async () => {
    const [exitCode, output] = await runCli('test/fixtures');

    expect(exitCode).toBe(1);
    expect(output).toMatchSnapshot();
  });

  test('should accept an ignore glob', async () => {
    const [exitCode, output] = await runCli('test/fixtures', '--ignore', '**/*.soy');

    expect(exitCode).toBe(0);
    expect(output).toMatchSnapshot();
  });

  test('should run using config file', async () => {
    const [exitCode, output] = await runCliFrom(
      '../../../lib/index.js',
       './test/fixtures/config',
      '../TransformedImport.soy');

    expect(exitCode).toBe(0);
    expect(output).toMatchSnapshot();
  });

  test('should fail with invalid config file', async () => {
    const [exitCode, output] = await runCliFrom(
      '../../../lib/index.js',
       './test/fixtures/invalid-config',
      '../TransformedImport.soy');

    expect(exitCode).toBe(1);
    expect(output).toMatchSnapshot();
  });
});
