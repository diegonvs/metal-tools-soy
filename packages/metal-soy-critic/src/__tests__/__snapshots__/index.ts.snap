// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should accept an ignore glob 1`] = `
"0 out of 0 file(s) have problems:

"
`;

exports[`cli should print more with verbose flag 1`] = `
"0 out of 1 file(s) have problems:

File - test/fixtures/Test.soy

     No problems were found.


"
`;

exports[`cli should print status when no errors 1`] = `
"0 out of 1 file(s) have problems:

"
`;

exports[`cli should print usage without args 1`] = `
"
  Usage: index mcritic [options] <path>

  Options:

    -h, --help             output usage information
    -V, --version          output the version number
    -i, --ignore <ignore>  A glob to ignore files, if passed a directory
    -v, --verbose          Output for all files

"
`;

exports[`cli should recursively look for Soy files if passed a directory 1`] = `
"18 out of 26 file(s) have problems:

File - test/fixtures/BadTernaryElvis.soy

     Use the Elvis (?:) operator instead of a ternary for default values:
     
     $optionalInfo - Line 20


File - test/fixtures/BadWhitespace.soy

     There seems to be extra whitespace on these lines:
     
     Line 8: {@param title: string }


File - test/fixtures/DelTemplate.soy

     Cannot find a .render template, do you have a typo?
     
     First
     Second


File - test/fixtures/Invalid.soy

     Expected: '-'
     '=\\"'
     a letter


File - test/fixtures/InvalidJS.soy

     Failed to parse component (javascript) file


File - test/fixtures/InvalidNamespace.soy

     Expected: '{namespace'


File - test/fixtures/LetStatements.soy

     These map keys should be sorted:
     
     Lines 15 to 16



     Use the Elvis (?:) operator instead of a ternary for default values:
     
     $name - Line 20


File - test/fixtures/MapParam.soy

     Please sort the params in these calls:
     
     OtherComponent.render - Lines 9 to 15



     These map keys should be sorted:
     
     Lines 11 to 13


File - test/fixtures/MissingCallImport.soy

     It looks like the following component calls are missing an import:
     
     MyExternalComponent


File - test/fixtures/MissingInternal.soy

     Based on their name, these attributes should have the .internal() config added:
     
     count_


File - test/fixtures/MissingParams.soy

     These params can't be found in STATE or your Class:
     
     optionalInfo


File - test/fixtures/MissingRender.soy

     Cannot find a .render template, do you have a typo?
     
     rander


File - test/fixtures/MissingRequired.soy

     These attributes are required in your Soy Template but not in your Component:
     
     title



     These attributes are .required() in your Component but not in your Template:
     
     description


File - test/fixtures/NestedCalls.soy

     Please sort the params in these calls:
     
     OtherComponent.render - Lines 28 to 32


File - test/fixtures/NoopEvents.soy

     These calls are not components, so their events param will do nothing:
     
     OtherComponent.Helper


File - test/fixtures/SortedParamDeclarations.soy

     These templates need their param declarations sorted:
     
     .render


File - test/fixtures/SortedParams.soy

     Please sort the params in these calls:
     
     OtherComponent.render - Lines 8 to 11


File - test/fixtures/UnsortedMap.soy

     These map keys should be sorted:
     
     Lines 10 to 12
     Lines 18 to 20


"
`;

exports[`cli should validate file 1`] = `
"1 out of 1 file(s) have problems:

File - test/fixtures/MissingInternal.soy

     Based on their name, these attributes should have the .internal() config added:
     
     count_


"
`;
